compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -290,12 +292,43 @@ getTypeClassDictionary world = \case
     _ ->
         Nothing
 
-simplifyExpr :: World -> Expr -> Expr
-simplifyExpr world = fst . cata go
+simplifyExpr :: Expr -> Simplifier Expr
+simplifyExpr = fmap fst . cata go'
   where
-
-    go :: ExprF (Expr, Info) -> (Expr, Info)
-    go = \case
+    go' :: ExprF (Simplifier (Expr, Info)) -> Simplifier (Expr, Info)
+    go' ms = do
+        es <- sequence ms
+        world <- gets sWorldExtended
+        let v' = freeVarsStep (fmap (freeVars . snd) es)
+        if freeVarsNull v'
+          then pure (go world es)
+          else do -- constant lifting
```

Comment:
Why do we not do constant lifting if the list of free variables is null?

Reply:
The idea was to group together the constant liftings, so instead of doing one lift per application (for example), we will lift all the applications at once, as far as we can. This also prevents us from doing a constant lift in cases where the binding is already constant...

(It might be better to also check isWorthLifting ... I think I'll experiment with that.)

Reply:
Oh I see that makes sense. Would be great to add a comment. Thanks for the explanation!